### DISCUSSION TITLE
seems some incosistent with MG_ARCH_CUSTOM and mongoose_custom.h

### DISCUSSION THREAD

**Discussion by nikitos1550** [cesanta/mongoose#1825] [Ask questions, get answers]
- My goal is: build mongoose with MG_ARCH_CUSTOM
- My actions were: compile mongoose.h
- My expectation was: successful compilation
- The result I saw: error
- My question is: What is proper way to build mongoose for custom arch

# Environment

- mongoose version: 7.8
- Compiler/IDE and SDK: 
```
arm-none-eabi-gcc -v
Using built-in specs.
COLLECT_GCC=arm-none-eabi-gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-none-eabi/6.3.1/lto-wrapper
Target: arm-none-eabi
Configured with: ../src/configure --build=x86_64-linux-gnu --prefix=/usr --includedir='/usr/lib/include' --mandir='/usr/lib/share/man' --infodir='/usr/lib/share/info' --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir='/usr/lib/lib/x86_64-linux-gnu' --libexecdir='/usr/lib/lib/x86_64-linux-gnu' --disable-maintainer-mode --disable-dependency-tracking --mandir=/usr/share/man --enable-languages=c,c++,lto --enable-multilib --disable-decimal-float --disable-libffi --disable-libgomp --disable-libmudflap --disable-libquadmath --disable-libssp --disable-libstdcxx-pch --disable-nls --disable-shared --disable-threads --disable-tls --build=x86_64-linux-gnu --target=arm-none-eabi --with-system-zlib --with-gnu-as --with-gnu-ld --with-pkgversion=15:6.3.1+svn253039-1build1 --without-included-gettext --prefix=/usr/lib --infodir=/usr/share/doc/gcc-arm-none-eabi/info --htmldir=/usr/share/doc/gcc-arm-none-eabi/html --pdfdir=/usr/share/doc/gcc-arm-none-eabi/pdf --bindir=/usr/bin --libexecdir=/usr/lib --libdir=/usr/lib --disable-libstdc++-v3 --host=x86_64-linux-gnu --with-headers=no --without-newlib --with-multilib-list=rmprofile CFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' CPPFLAGS='-Wdate-time -D_FORTIFY_SOURCE=2' CXXFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' FCFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' FFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' GCJFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' LDFLAGS='-Wl,-Bsymbolic-functions -Wl,-z,relro' OBJCFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' OBJCXXFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-iopiMw/gcc-arm-none-eabi-6.3.1+svn253039=. -fstack-protector-strong' INHIBIT_LIBC_CFLAGS=-DUSE_TM_CLONE_REGISTRY=0 AR_FOR_TARGET=arm-none-eabi-ar AS_FOR_TARGET=arm-none-eabi-as LD_FOR_TARGET=arm-none-eabi-ld NM_FOR_TARGET=arm-none-eabi-nm OBJDUMP_FOR_TARGET=arm-none-eabi-objdump RANLIB_FOR_TARGET=arm-none-eabi-ranlib READELF_FOR_TARGET=arm-none-eabi-readelf STRIP_FOR_TARGET=arm-none-eabi-strip
Thread model: single
gcc version 6.3.1 20170620 (15:6.3.1+svn253039-1build1)
```
- Target hardware/board: wch ch579 evt 
- Connectivity chip/module: ch579

I am following https://mongoose.ws/documentation/#custom-build
I added compile flag -DMG_ARCH=MG_ARCH_CUSTOM
I created `mongoose_custom.h`:
```
#include <stdint.h>
#include <stdbool.h>
#include <stdarg.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>

#define MG_DIRSEP '/'
#define MG_INT64_FMT "%lld"

#define MG_ENABLE_SOCKET 0
#define MG_ENABLE_SSI 0
```

From mongoose.h
```
#if !defined(MG_ARCH)
#if defined(__unix__) || defined(__APPLE__)
#define MG_ARCH MG_ARCH_UNIX
#elif defined(_WIN32)
#define MG_ARCH MG_ARCH_WIN32
#elif defined(ICACHE_FLASH) || defined(ICACHE_RAM_ATTR)
#define MG_ARCH MG_ARCH_ESP8266
#elif defined(__ZEPHYR__)
#define MG_ARCH MG_ARCH_ZEPHYR
#elif defined(ESP_PLATFORM)
#define MG_ARCH MG_ARCH_ESP32
#elif defined(FREERTOS_IP_H)
#define MG_ARCH MG_ARCH_FREERTOS_TCP
#elif defined(AZURE_RTOS_THREADX)
#define MG_ARCH MG_ARCH_AZURERTOS
#elif defined(PICO_TARGET_NAME)
#define MG_ARCH MG_ARCH_RP2040
#endif

#if !defined(MG_ARCH)
#include "mongoose_custom.h"  // keep this include
#endif

#if !defined(MG_ARCH)
#error "MG_ARCH is not specified and we couldn't guess it. Set -D MG_ARCH=..."
#endif
#endif  // !defined(MG_ARCH)
```
Seems if I will setup MG_ARCH=MG_ARCH_CUSTOM then `mongoose_custom.h` never will be triggered, that is different (as I understood) from the manual explanation. 

My final target to startup mongoose web server in ch579 (a bit info here https://github.com/SoCXin/CH579). This chip has intergrated 10M ethernet phy and vendor provides closed source library (as static archive) with some api that can be used as underlayer for mongoose (at the moment I hope so).

**Comment by scaprile** [comment]
It looks like you are right. I will take a look at it on Monday, in the meantime you can watch a video tutorial or read a tutorial where we do a custom build for STM32. Do not define MG_ARCH and the custom header file will be included, that's for sure.
https://www.youtube.com/watch?v=8htC_TSBeO0
https://mongoose.ws/tutorials/stm32/nucleo-f746zg-cube/
As I see some "arm" and "gcc" in there, your case shouldn't be very different from the STM32 case described in the tutorials
If you have more questions, we can continue here.

**Reply by nikitos1550** [reply]
Yes, I followed second link tutorial and now I can compile mongoose!
What is MG_ARCH_NEWLIB? Is it somehow related to --specs=nano.specs ?

**Reply by scaprile** [reply]
Yes, that is a C library for embedded systems and nano is great for bare metal on (ARM) microcontrollers
https://mongoose.ws/documentation/#build-options
Basic C built-ins in a reduced size.
https://sourceware.org/newlib/
https://stackoverflow.com/questions/65453293/what-are-nosys-nano-rdimon-terms-when-using-arm-gcc
https://community.arm.com/arm-community-blogs/b/embedded-blog/posts/shrink-your-mcu-code-size-with-gcc-arm-embedded-4-7

**Reply by nikitos1550** [reply]
If I use MG_ARCH_NEWLIB, how can I bind network operations to the closed lib from vendor? Still a bit not clear.

**Reply by scaprile** [reply]
Well, when you are in "custom mode" you need to provide some sort of abstraction layer. Mongoose itself uses BSD sockets calls, you have to write something like what is in the arch_x.h files in the src/ directory https://github.com/cesanta/mongoose/tree/master/src
If your vendor provides the usual socket() recv() send() select() you won't have any issues. Otherwise, you have to either provide that layer or maybe MIP is a better option, where you use Mongoose's TCP/IP stack and you need to provide a driver for layer-2 input/output (Ethernet frames) https://github.com/cesanta/mongoose/tree/master/mip
The video and tutorial linked above use MIP.

**Reply by nikitos1550** [reply]
I inspected vendor lib, it also uses sockets, but it is different to BSD sockets API, so some layer should be between mongoose and vendor's api. What functions should I implement?
https://github.com/cesanta/mongoose/blob/master/test/mongoose_custom.c this list 
or something like this list:
```
// Why FreeRTOS-TCP did not implement a clean BSD API, but its own thing
// with FreeRTOS_ prefix, is beyond me
#define IPPROTO_TCP FREERTOS_IPPROTO_TCP
#define IPPROTO_UDP FREERTOS_IPPROTO_UDP
#define AF_INET FREERTOS_AF_INET
#define SOCK_STREAM FREERTOS_SOCK_STREAM
#define SOCK_DGRAM FREERTOS_SOCK_DGRAM
#define SO_BROADCAST 0
#define SO_ERROR 0
#define SOL_SOCKET 0
#define SO_REUSEADDR 0
#define sockaddr_in freertos_sockaddr
#define sockaddr freertos_sockaddr
#define accept(a, b, c) FreeRTOS_accept((a), (b), (c))
#define connect(a, b, c) FreeRTOS_connect((a), (b), (c))
#define bind(a, b, c) FreeRTOS_bind((a), (b), (c))
#define listen(a, b) FreeRTOS_listen((a), (b))
#define socket(a, b, c) FreeRTOS_socket((a), (b), (c))
#define send(a, b, c, d) FreeRTOS_send((a), (b), (c), (d))
#define recv(a, b, c, d) FreeRTOS_recv((a), (b), (c), (d))
#define setsockopt(a, b, c, d, e) FreeRTOS_setsockopt((a), (b), (c), (d), (e))
#define sendto(a, b, c, d, e, f) FreeRTOS_sendto((a), (b), (c), (d), (e), (f))
#define recvfrom(a, b, c, d, e, f) \
  FreeRTOS_recvfrom((a), (b), (c), (d), (e), (f))
#define closesocket(x) FreeRTOS_closesocket(x)
#define gethostbyname(x) FreeRTOS_gethostbyname(x)
#define getsockname(a, b, c) (-1)
#define getpeername(a, b, c) 0
```

By the way, now mongoose built ok, but I didn't provide any functions, I expected that there will be errors, that some functions doesn't exist.

**Reply by nikitos1550** [reply]
P.S. MIP is not applicable here, as there is no documentation on ethernet IP block and vendor will never give any low level information, so such layer-2 driver is possible to do via reverse engineering proprietary lib, that for sure is not what I am gonna do.

**Reply by scaprile** [reply]
The stuff in test/ is usually for testing (with some exceptions...)
You must provide the BSD socket functions, flags, and structures, that means: yes, something similar to those defines for the FreeRTOS architecture; actual functions if you can't get away with defines (e.g.: the functions use different parameters/order, return different values, etc.)
You should get link errors... if you don't, those functions are being provided by "someone" "somewhere", if you didn't get compiler warnings then you are providing headers for those functions and your compiler knows where to get them for linking... It is hard to follow without actually seeing your custom.h and Makefile files, and any compiler messages; etc...

**Comment by nikitos1550** [comment]
Here is my mongoose_custom.h file:
```
#ifndef MONGOOSE_CUSTOM_H
#define MONGOOSE_CUSTOM_H

#include <errno.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

//#define MG_ARCH MG_ARCH_NEWLIB
#define MG_ARCH MG_ARCH_CUSTOM

#define MG_DIRSEP '/'
#define MG_INT64_FMT "%lld"

#define MG_ENABLE_SOCKET 0
#define MG_ENABLE_SSI 0

#define MG_ENABLE_FILE 0
#define MG_ENABLE_PACKED_FS 0

#define MG_ENABLE_LINES 0

#define MG_ENABLE_LOG 1

#endif // MONGOOSE_CUSTOM_H
```

And with this includes, I don't have any linking issues, that is strange.

**Reply by scaprile** [reply]
> #define MG_ENABLE_SOCKET 0
> And with this includes, I don't have any linking issues, that is strange.

That's weird, you are not including any socket code, you should be getting link errors because at least mg_send() can't be found. Also mg_mgr_poll() will be missing so you can't actually do anything.

**Reply by nikitos1550** [reply]
If I setup `#define MG_ENABLE_SOCKET 0` then mongoose will try use ` mg_send()` and etc?
I also created mongoose_custom.c (as I followed some tutorial before):
```
#include "mongoose.h"

int mkdir(const char *path, mode_t mode) {
  (void)path, (void)mode;
  return -1;
}

void mg_connect_resolved(struct mg_connection *c) { (void)c; }

bool mg_open_listener(struct mg_connection *c, const char *url) {
  (void)c, (void)url;
  return false;
}

void mg_mgr_poll(struct mg_mgr *mgr, int ms) { (void)mgr, (void)ms; }

bool mg_send(struct mg_connection *c, const void *buf, size_t len) {
  (void)c, (void)buf, (void)len;
  return false;
}

int mg_mkpipe(struct mg_mgr *m, mg_event_handler_t fn, void *d, bool udp) {
  (void)m, (void)fn, (void)d, (void)udp;
  return -1;
}

// void _fini(void);
// void _fini(void) {
//}
```

But actually I don't know what these functions should do. Maybe there are some reference?
Am I right that there are two ways to bind mongoose on some platform sockets or these functions?

**Reply by scaprile** [reply]
Mongoose uses its internal functions (mg_send()...), those need to be implemented by some sort of a connection to a networking stack. Currently is sockets or MIP. So basically you define MG_ENABLE_SOCKET 1 and provide a BSD socket layer, or define MIP and use MIP.
Or... you port src/sock.c to your environment. Your tutorial seems to be trying to do the later, and it is probably way outdated.
The easiest path IMHO is to interface to your system in a way similar to what arch_FreeRTOS does

**Reply by scaprile** [reply]
You may skip the filesystem functions if you use our packed filesystem: https://mongoose.ws/tutorials/embedded-filesystem/, 
but for networking you either provide a standard access layer (linkable BSD socket functions) or a low-level driver to the Ethernet/WiFi controller.

**Reply by nikitos1550** [reply]
I see, ok, let me try the way you suggest. Will report here later. Thanks.

**Reply by cpq** [reply]
@nikitos1550 what TCP/IP stack are you using?
